src/qutip_qip/circuit/text_renderer.py

```diff
@@ -0,0 +1,497 @@
+
+__all__ = [
+    "TextRenderer",
+]
+
+
+class TextRenderer(BaseRenderer):
+    """
+    A class to render a quantum circuit in text format.
+    """
+
+    def __init__(self, qc: QubitCircuit, **style):
+
+        # user defined style
+        style = {} if style is None else style
+        style["gate_margin"] = 0
+        self.style = StyleConfig(**style)
+
+        super().__init__(self.style)
+        self._qc = qc
+        self._qwires = qc.N
+        self._cwires = qc.num_cbits
+        self._layer_list = {i: [] for i in range(self._qwires + self._cwires)}
+
+        self._render_strs = {
+            "top_frame": ["  "] * (self._qwires + self._cwires),
+            "mid_frame": ["──"] * self._qwires + ["══"] * self._cwires,
+            "bot_frame": ["  "] * (self._qwires + self._cwires),
+        }
+
+    def _adjust_layer(self, wire_list: List[int], xskip: int):
+        """
+        Adjust the layers by filling the empty spaces with respective characters
+        """
+
+        for wire in wire_list:
+            self._render_strs["top_frame"][wire] += " " * (
+                xskip - len(self._render_strs["top_frame"][wire])
+            )
+            self._render_strs["bot_frame"][wire] += " " * (
+                xskip - len(self._render_strs["bot_frame"][wire])
+            )
+            if wire < self._qwires:
+                self._render_strs["mid_frame"][wire] += "─" * (
+                    xskip - len(self._render_strs["mid_frame"][wire])
+                )
+            else:
+                self._render_strs["mid_frame"][wire] += "═" * (
+                    xskip - len(self._render_strs["mid_frame"][wire])
+                )
+
+    def _add_wire_labels(self):
+        """
+        Add wire labels to the circuit
+        """
+
+        default_labels = [f"q{i}" for i in range(self._qwires)] + [
+            f"c{i}" for i in range(self._cwires)
+        ]
+        for i, label in enumerate(default_labels):
+            self._render_strs["mid_frame"][i] = (
+                f" {label} :" + self._render_strs["mid_frame"][i]
+            )
+            self._render_strs["top_frame"][i] = " " * (len(f" {label} :") + 2)
+            self._render_strs["bot_frame"][i] = " " * (len(f" {label} :") + 2)
+
+        self._layer_list = {
+            i: [len(self._render_strs["mid_frame"][i])]
+            for i in range(self._qwires + self._cwires)
+        }
+
+    def _extend_line(self):
+        """
+        Extend all the wires to the same length
+        """
+
+        wire_len = [sum(wire) for wire in self._layer_list.values()]
+        max_len = max(wire_len) + self.style.end_wire_ext
+
+        for i, length in enumerate(wire_len):
+            if length < max_len:
+                diff = max_len - length
+                if i < self._qwires:
+                    self._render_strs["mid_frame"][i] += "─" * diff
+                else:
+                    self._render_strs["mid_frame"][i] += "═" * diff
```

Comment:
What is the difference between `_extend_line` and `_adjust_layer`. Why does `_extend_line` not add space to the top and bottom? I feel like `_extend_line` can be replaced by `adjust_layer` by a proper definition of `xskip`

Reply:
- Since we're dealing with strings and the gates vary in width, `_adjust_layer` adds padding to maintain consistent layer alignment across the different wires. 

- `extend_wire` is only used at the end of circuit, after all gates are added to extend wire to certain extra length

Reply:
It's a nice idea to reuse the `_adjust_layer` instead of `_extend_wire`, will give it a try.

Reply:
Update: Removed `_extend_line`